core/solver/trs.cpp

```diff
@@ -0,0 +1,112 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2019, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/trs.hpp>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/solver/trs_kernels.hpp"
+
+namespace gko {
+namespace solver {
+
+
+namespace trs {
+
+
+GKO_REGISTER_OPERATION(generate, trs::generate);
+GKO_REGISTER_OPERATION(solve, trs::solve);
+
+
+}  // namespace trs
+
+
+template <typename ValueType, typename IndexType>
+void Trs<ValueType, IndexType>::generate(const LinOp *system_matrix,
+                                         const LinOp *b)
+{
+    using CsrMatrix = matrix::Csr<ValueType, IndexType>;
+    using Vector = matrix::Dense<ValueType>;
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+    // This is needed because it does not make sense to call the copy and
+    // convert if the existing matrix (if not CSR) is empty.
+    if (dynamic_cast<const CsrMatrix *>(system_matrix) == nullptr) {
+        GKO_ASSERT_IS_NON_EMPTY_MATRIX(system_matrix);
```

Comment:
By the way, as far as I can tell, this would be the first time we test if the matrix is empty. I actually don't know what would happen if we would run other solvers (or this solver without this check) with an empty matrix.
Did you check what would happen without it?

Also: In the current form, you would need a test for the inside of this `if`. Either way, I would reccommend a test to see if (inside `Trs::generate`) the proper error is thrown in case an empty matrix is given.

Reply:
All solvers can be created with an empty matrix. `trs` and possibly the `ilu` factorization, in which the actual kernel operates on a CSR matrix, can be created with an empty CSR matrix but not with other matrix formats as empty formats are not convertible (I think).